### PR TITLE
Internal: Support additional types validation on save [EDS-444]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php [hunk 1]
```diff
@@ -42,25 +50,33 @@ public function validate_value( $value ): void {
...
 			throw new \Exception( "Dynamic tag `{$value['name']}` does not exist" );
 		}
 
 		if ( ! $this->is_tag_in_supported_categories( $tag ) ) {
-			$tag_categories = implode( ', ', $tag->get_categories() );
+			$tag_categories = implode( ', ', $tag['categories'] );
 			$prop_categories = implode( ', ', $this->get_categories() );
 
-			throw new \Exception( "Dynamic tag `{$tag->get_name()}` categories ($tag_categories) are not in supported categories ($prop_categories)" );
+			throw new \Exception( "Dynamic tag `{$tag['name']}` categories ($tag_categories) are not in supported categories ($prop_categories)" );
```
Comment (Nevoss) [inline modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php:63]: Extract into fn

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php [hunk 2]
```diff
@@ -1,16 +1,24 @@
 <?php
```
Comment (Nevoss) [inline modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php:1]: Move the transformer

File: modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		return static::$instance;
+	}
+
+	public static function fresh(): self {
+		return static::instance( true );
+	}
+
+	public function register_hooks() {
+		add_action( 'elementor/init', function () {
+			$v1_tags = Plugin::$instance->dynamic_tags->get_tags_config();
```
Comment (Nevoss) [inline modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php:31]: On-demand and not on event

File: modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+			$this->registry->populate_from_v1_tags( $v1_tags );
```
Comment (Nevoss) [inline modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php:33]:
```suggestion
```

File: modules/atomic-widgets/settings-validator.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+			$validated[ $key ] = $value;
+		}
+
+		$is_valid = empty( $errors );
+
+		return [
+			$is_valid,
+			$validated,
+			$errors,
+		];
```
Comment (Nevoss) [inline modules/atomic-widgets/settings-validator.php:62]:
```suggestion
		return Validation_Result::make();
```

Comment (Nevoss) [reply modules/atomic-widgets/settings-validator.php:62]: maybe?

File: modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+			fn( array $settings ) => $this->add_atomic_dynamic_tags_settings( $settings )
+		);
+
+		add_filter(
+			'elementor/atomic-widgets/props-schema',
+			fn( array $schema ) => Dynamic_Prop_Types_Mapping::add_to_schema( $schema )
+		);
+	}
+
+	private function add_atomic_dynamic_tags_settings( $settings ) {
```
Comment (Nevoss) [inline modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php:47]:
```suggestion
	private function add_atomic_dynamic_tags_to_editor_settings( $settings ) {
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: modules/atomic-widgets/base/atomic-widget-base.php [hunk 7]
```diff
@@ -149,26 +146,16 @@ private static function validate_schema( array $schema ) {
...
-
-					if ( null !== $sanitized_value ) {
-						$sanitized_values[ $key ] = $sanitized_value;
-					}
-				} catch ( \Exception $e ) {
-					Utils::safe_throw( "Error while sanitizing `$key` prop in `{$widget_name}` - {$e->getMessage()}" );
-				}
-			}
+		if ( ! empty( $errors ) ) {
+			Utils::safe_throw( 'Settings validation failed. Invalid keys: ' . join( ', ', $errors ) );
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply modules/atomic-widgets/base/atomic-widget-base.php:155]: Done

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php (hunk 1 above)
Comment (StyleShit) [reply modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php:63]: Done

File: modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php (hunk 2 above)
Comment (StyleShit) [reply modules/atomic-widgets/dynamic-tags/dynamic-prop-type.php:1]: Done

File: modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php (hunk 3 above)
Comment (StyleShit) [reply modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php:31]: Done

File: modules/atomic-widgets/settings-validator.php (hunk 5 above)
Comment (StyleShit) [reply modules/atomic-widgets/settings-validator.php:62]: no

File: modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php (hunk 6 above)
Comment (StyleShit) [reply modules/atomic-widgets/dynamic-tags/dynamic-tags-module.php:47]: Done

**Review by Nevoss — APPROVED** [review]

**Pull request merged**